pkg/auth/github.go

```diff
@@ -0,0 +1,158 @@
+package auth
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+
+	"github.com/fastschema/fastschema/fs"
+
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/github"
+)
+
+const GithubLoginOauthURL = "https://github.com/login/oauth/access_token"
+const GithubUserURL = "https://api.github.com/user"
+
+type GithubAccessTokenResponse struct {
+	Scope       string `json:"scope"`
+	TokenType   string `json:"token_type"`
+	AccessToken string `json:"access_token"`
+}
+
+type GithubUserResponse struct {
+	Login     string `json:"login"`
+	ID        int    `json:"id"`
+	AvatarURL string `json:"avatar_url"`
+	Name      string `json:"name"`
+	Blog      string `json:"blog"`
+	Email     string `json:"email"`
+	Bio       string `json:"bio"`
+}
+
+type GithubAuthProvider struct {
+	oauth *oauth2.Config
+}
+
+func NewGithubAuthProvider(
+	config map[string]string,
+	redirectURL string,
+) (fs.AuthProvider, error) {
+	if config["client_id"] == "" || config["client_secret"] == "" {
+		return nil, fmt.Errorf("github client id or secret is not set")
+	}
+
+	return &GithubAuthProvider{
+		oauth: &oauth2.Config{
+			ClientID:     config["client_id"],
+			ClientSecret: config["client_secret"],
+			RedirectURL:  redirectURL,
+			Endpoint:     github.Endpoint,
+		},
+	}, nil
+}
+
+func (ga *GithubAuthProvider) Name() string {
+	return "github"
+}
+
+func (ga *GithubAuthProvider) Login(c fs.Context) (_ any, err error) {
+	rediectURL := ga.oauth.AuthCodeURL(
+		c.Cookies("UUID"),
```

Comment:
The state is a random string to support CSRF protection. We need mechanism to generate and verify the state value. Use the cookie value seem unsafe, so we will skip this for now.

```suggestion
		// use random string
```